src/id_allocator.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+// Error type for IdAllocator usage.
+#[derive(Debug)]
+pub enum Error {
+    // All ids from the range specified are exhausted.
+    Overflow,
+    // An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
```

Comment:
If this is only used for out of range IDs, can we rename this to `OutOfRange` or similar so that the error variant is more verbose?

Reply:
It is not used only for out of range IDs. This is also returned when we try to release an ID that was never allocated or already released.

Reply:
Can we have another error for that one instead? `InvalidInput` is a bit too generic because most of errors are caused by an invalid input in the end.